src/Analyser/MutatingScope.php

```diff
@@ -3758,51 +3721,20 @@ public function filterBySpecifiedTypes(SpecifiedTypes $specifiedTypes): self
 			} else {
 				$scope = $scope->removeTypeFromExpression($expr, $type);
 			}
-
-			if (
-				!$expr instanceof Variable
-				|| !is_string($expr->name)
-				|| $specifiedTypes->shouldOverwrite()
-			) {
-				// @phpstan-ignore-next-line
-				$match = Strings::match((string) $typeSpecification['exprString'], '#^(\$[a-zA-Z_\x7f-\xff][a-zA-Z_0-9\x7f-\xff]*)#');
-				if ($match !== null) {
-					$skipVariables[$match[1]] = true;
-				}
-				continue;
-			}
-
-			if ($scope->hasVariableType($expr->name)->no()) {
-				continue;
-			}
-
-			$saveConditionalVariables['$' . $expr->name] = $scope->getVariableType($expr->name);
-		}
-
-		foreach ($saveConditionalVariables as $variableExprString => $typeGuard) {
-			if (array_key_exists($variableExprString, $skipVariables)) {
-				continue;
-			}
-
-			$typeGuards[$variableExprString] = $typeGuard;
 		}
 
 		$newConditionalExpressions = $specifiedTypes->getNewConditionalExpressionHolders();
 		foreach ($this->conditionalExpressions as $variableExprString => $conditionalExpressions) {
-			if (array_key_exists($variableExprString, $typeGuards)) {
-				continue;
-			}
-
 			$typeHolder = null;
 
 			foreach ($conditionalExpressions as $conditionalExpression) {
 				$matchingConditions = [];
 				foreach ($conditionalExpression->getConditionExpressionTypes() as $conditionExprString => $conditionalType) {
-					if (!array_key_exists($conditionExprString, $typeGuards)) {
+					$conditionExpr = $this->exprStringToExpr($conditionExprString);
+					if ($conditionExpr === null) {
 						continue;
 					}
-
-					if (!$typeGuards[$conditionExprString]->equals($conditionalType)) {
+					if (!$scope->getType($conditionExpr)->equals($conditionalType)) {
```

Comment:
This line is causing a infinite recursion, not sure why yet.

Reply:
recursion with this line
https://github.com/rajyan/phpstan-src/blob/0ae0fe306679e400843bcf2c6de36f67be1eaa9b/src/Analyser/MutatingScope.php#L1601-L1603

Reply:
Solved